### PR TITLE
fix(Core/Spells): Category cooldowns should be processed only within …

### LINKED ISSUE #5475: [Spell] Rank 1 Flamestrike Shares Cooldown With All Explosives [$5 awarded]
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [X] enUS
- [ ] other (specify)

##### FACTION

- [X] Alliance
- [X] Horde

##### CONTENT PHASE:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

##### CURRENT BEHAVIOUR:

Original report: https://github.com/chromiecraft/chromiecraft/issues/466

When a dynamite/bomb is used, flamestrike will share a cooldown with them (1min)
The cd is not shown on flamestrike;
You'll see "Spell is not ready yet"

When I tested this, I tried all 12 bombs available to engineers at lvl 225 with every rank of Flamestrike. Only rank 1 Flamestrike was affected. All 12 bombs caused the 1 min cooldown, even Explosive Sheep.

WoWScrnShot_042621_125141

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Flamestrike should not share a cd with these explosives.

##### STEPS TO REPRODUCE THE PROBLEM:

1. Use a dynamite/bomb from Engineering
2. Use Flamestrike rank 1.

##### EXTRA NOTES:

Bombs tried:
[Mithril Frag Bomb]
[Big Iron Bomb] 
[Big Bronze Bomb] 
[Small Bronze Bomb] 
[Large Copper Bomb] 
[Rough Copper Bomb]
[Solid Dynamite] 
[Heavy Dynamite] 
[Coarse Dynamite]
[Ez-Thro Dynamite] 
[Rough Dynamite]
[Explosive Sheep]

https://wowgaming.altervista.org/aowow/?spell=2120

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/91081f4ad8fed98d7bdc9c0ae13d05c86032df9f

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

<bountysource-plugin>

---
The **[$5 bounty](https://www.bountysource.com/issues/98101600-spell-rank-1-flamestrike-shares-cooldown-with-all-explosives?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue has been claimed at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
…the same spellfamily.

Fixed #5475.

## Changes Proposed:
-  Spell cooldowns came from categories should be handled only if the triggering spell is the same family.

## Issues Addressed:

- Closes #5475
- Closes https://github.com/chromiecraft/chromiecraft/issues/466
## Tests Performed:

- Not tested.

## How to Test the Changes:

1.Use a dynamite/bomb from Engineering
2. Use Flamestrike rank 1.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -23085,6 +23085,11 @@ void Player::AddSpellAndCategoryCooldowns(SpellInfo const* spellInfo, uint32 ite
                 if (*i_scset == spellInfo->Id)                    // skip main spell, already handled above
                     continue;
 
+                // Only within the same spellfamily
+                SpellInfo const* categorySpellInfo = sSpellMgr->GetSpellInfo(*i_scset);
+                if (!categorySpellInfo || categorySpellInfo->SpellFamilyName != spellInfo->SpellFamilyName)
+                    continue;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp:23091]:
```suggestion
                if (!categorySpellInfo || categorySpellInfo->SpellFamilyName != spellInfo->SpellFamilyName)
                {
                    continue;
                }
```

**Comment by UltraNix** [comment]
Please, test it.

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @mpfans @pklloveyou @Eldarix  can anyone please test this?

**Comment by tkn963** [comment]
I'm just saying this to give a heads up but I'm currently compiling the PR and will be testing it soon. Going to be using several different types of explosives and/or bombs from engineering and Flamestrike. Will give a list of engineering items used and the results when I'm done.

**Comment by tkn963** [comment]
Tested the following items followed by using Flamestrike (rank 1 and 9 just to be inclusive)

Explosive Decoy (Item ID: 40536)
Explosive Sheep (Item ID: 4384)
Cobalt Frag Bomb (Item ID: 40771)
Arcane Bomb (Item ID: 16040)
Dark Iron Bomb (Item ID: 16005)
Dense Dynamite (Item ID: 18641)
Solid Dynamite (Item ID: 10507)
Ez-Thro Dynamite (Item ID: 6714)
Coarse Dynamite (Item ID: 4365)
Rough Dynamite (Item ID: 4358)

No issues! 👍

**Comment by Sedatyf** [comment]
Awesome testing work @tkn963! Thanks a lot!
Thanks @UltraNix for this PR
🚀

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
good jov @UltraNix and thank you @tkn963 for your excellent testing feedback!

**Comment by necropola** [comment]
@UltraNix Can you please add

- Closes #6324
- Closes https://github.com/chromiecraft/chromiecraft/issues/845

Just tested your PR and it solves the **Lesser Heal** issue, too.

**Pull request merged**